File: original_scripts/c++/binary_search/deepseek-coder_6.7b.cpp
```cpp
template<typename T>
int binary_search(T arr[], int start, int end, T target) {
    if (end >= start) {
        int mid = start + (end - start) / 2;
        
        if (arr[mid] == target) return mid;
  
        if (arr[mid] > target) return binary_search(arr, start, mid - 1, target);
  
        return binary_search(arr, mid + 1, end, target);
    }
    
    return -1;
}
```

File: original_scripts/rust/binary_search/deepseek-coder_6.7b.rs
```rust
pub fn binary_search<T: Ord>(arr: &[T], target: T) -> Option<usize> {
    let mut start = 0;
    let mut end = arr.len();
    
    while start < end {
        let mid = (start + end) / 2;
        
        match arr[mid].cmp(&target) {
            std::cmp::Ordering::Equal => return Some(mid),
            std::cmp::Ordering::Less => start = mid + 1,
            std::cmp::Ordering::Greater => end = mid,
        }
    }
    
    None
}
```